rfc/001-truncated-block-history.md

```diff
@@ -0,0 +1,86 @@
+
+* [Block pruning](https://github.com/tendermint/tendermint/issues/3652): removes historical blocks and associated data (e.g. validator sets) up to some height, keeping only the most recent blocks.
+
+* [State sync](https://github.com/tendermint/tendermint/issues/828): bootstraps a new node by syncing state machine snapshots (note: not IAVL pruning snapshots) at a given height, but not historical blocks and associated data.
+
+To maintain the integrity of the chain, the use of these features must be coordinated such that necessary historical blocks will not become unavailable or lost forever. In particular:
+
+* Some nodes should have complete block histories, for auditability, querying, and bootstrapping.
+
+* The majority of nodes should retain blocks longer than the Cosmos SDK unbonding period, for light client verification.
+
+* Some nodes must take and serve state sync snapshots with snapshot intervals less than the block retention periods, to allow new nodes to state sync and then replay blocks to catch up.
+
+* Only a minority of nodes can be state synced within the unbonding period, for light client verification and to serve block histories for catch-up.
+
+However, it is unclear if and how we should enforce this. It may not be possible to technically enforce all of these without knowing the state of the entire network, but it may also be unrealistic to expect this to be enforced entirely through social coordination. This is especially unfortunate since the consequences of misconfiguration can be permanent chain-wide data loss.
+
+The main configuration options involved are:
+
+* Block retention (Tendermint): the number of recent blocks (heights) to retain.
+
+* Block time (Tendermint): the minimum duration between consecutive blocks.
+
+* Unbonding time (SDK): the duration in which validators can be economically punished for misbehavior.
+
+* Snapshot interval (SDK): the interval (in heights) between taking state sync snapshots.
+
+* Snapshot retention (SDK): the number of recent state sync snapshots to retain.
+
+## Proposal
+
+* Block retention (Tendermint): implement as a new consensus parameter `block.retention` (default 0, i.e. retain all), and add a local config option `consensus.prune_blocks` (default off) that allows operators to enable or disable block pruning on a per-node basis. Must be at minimum 2, since the current and previous blocks are required for progress.
+
+* Block time (Tendermint): already implemented as a consensus parameter `block.time_iota_ms`.
+
+* Unbonding time (SDK): already implemented as a genesis parameter `app_state.staking.params.unbonding_time`. Must now satisfy `unbonding_time < 2 * block.retention * block.time_iota_ms` otherwise errors (factor of 2 since `block.time_iota_ms` does not take into account block execution time).
```

Comment:
nit: maybe the `SDK` sections can be generalized for an application. We shouldn't have things that are needed specifically in the Cosmos-SDK in the Tendermint spec.

`An application building on Tendermint that would like to take advantage of these features should have x and y`

Reply:
I agree, but the problems we're discussing here are specific to the SDK. None of these would apply to e.g. the `kvstore` app, because it doesn't have a concept of unbonding time or snapshots.

Reply:
I'll try to generalize it a bit and/or move it to the SDK when we figure out what to do here, but I think it's good to keep the discussion as concrete as possible until then to avoid confusion.